src/Compilers/Core/Portable/SourceGeneration/Nodes/SyntaxTransformNode.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Text;
+using System.Threading;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal interface ISyntaxTransformNode
+    {
+        ISyntaxTransformBuilder GetBuilder(DriverStateTable previousStateTable);
+    }
+
+    internal interface ISyntaxTransformBuilder
+    {
+        bool Filter(SyntaxNode node);
+
+        void AddFilterFromPreviousTable(SemanticModel model, EntryState state);
+
+        void AddFilterEntries(ImmutableArray<SyntaxNode> nodes, SemanticModel model);
+
+        void SetInputState(DriverStateTable.Builder builder);
+    }
+
+    internal sealed class SyntaxTransformNode<T> : IIncrementalGeneratorNode<T>, ISyntaxTransformNode
+    {
+        private readonly Func<GeneratorSyntaxContext, T> _func;
+        private readonly IEqualityComparer<T> _comparer;
+        private readonly Func<SyntaxNode, bool> _filterFunc;
+
+        private readonly InputNode<SyntaxNode> _filterNode = new InputNode<SyntaxNode>();
+
+        internal SyntaxTransformNode(Func<SyntaxNode, bool> filterFunc, Func<GeneratorSyntaxContext, T> transformFunc, IEqualityComparer<T>? comparer = null)
+        {
+            _func = transformFunc;
+            _filterFunc = filterFunc;
+            _comparer = comparer ?? EqualityComparer<T>.Default;
+        }
+
+        // this is an input node, so we don't perform any updates
+        public NodeStateTable<T> UpdateStateTable(DriverStateTable.Builder graphState, NodeStateTable<T> previousTable, CancellationToken cancellationToken) => previousTable;
+
+        public IIncrementalGeneratorNode<T> WithComparer(IEqualityComparer<T> comparer) => new SyntaxTransformNode<T>(_filterFunc, _func, comparer);
+
+        public ISyntaxTransformBuilder GetBuilder(DriverStateTable previousStateTable) => new Builder(this, previousStateTable.GetStateTable(this), previousStateTable.GetStateTable(this._filterNode));
+
+        private sealed class Builder : ISyntaxTransformBuilder
+        {
+            private readonly SyntaxTransformNode<T> _owner;
+
+            private readonly NodeStateTable<T> _previousTransformTable;
+            private readonly NodeStateTable<T>.Builder _transformTable;
+
+            private readonly NodeStateTable<SyntaxNode> _previousFilterTable;
+            private readonly NodeStateTable<SyntaxNode>.Builder _filterTable;
+
+            public Builder(SyntaxTransformNode<T> owner, NodeStateTable<T> previousTransform, NodeStateTable<SyntaxNode> previousFilter)
```

Comment:
Consider reordering the last two parameters. Think the API is easier to use if we always keep the same order of 'filter' parameter then 'transform' parameter